### PR TITLE
Fix auth followup / Safe4337Base

### PR DESCRIPTION
**Pull request by voltrevo**
# *Dependent PR*

This PR depends on #120, please merge that PR first.

## Description

Followup for #118.

Changes:
- Override `_requireFromEntryPoint()` to use `_msgSender()` (comment)
  - Allows removal of `validateUserOp` overrides because the `BaseAccount` version now does the correct thing
- Prefer only allowing calls from entry point (comment)
- Implements the above as a common base class `Safe4337Base` which also allows deduplicating:
  - `_payPrefund()`
  - `_validateNonce()`
- `Safe4337Base` also implements `_currentSafe()` which helps clear up confusion for referring to the safe

### CODE REVIEW

File: account-integrations/safe/src/utils/Safe4337Base.sol [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+     * Get the current safe.
+     *
+     * This can be a bit counter-intuitive. The safe only does a delegatecall
+     * into the plugin during setup. In regular usage, it uses a regular call,
+     * which is why we can use `msg.sender` to refer to the safe. Additionally,
+     * `this` will be the plugin for those calls, so it won't work when trying
+     * to do safe operations like `execTransactionFromModule`.
+     */
+    function _currentSafe() internal virtual view returns (ISafe) {
+        return ISafe(msg.sender);
```
Comment (JohnGuilding) [inline account-integrations/safe/src/utils/Safe4337Base.sol:86]: Based on what was used before from the [safe example](https://github.com/safe-global/safe-contracts/blob/810fad9a074837e1247ca24ac9e7f77a5dffce19/contracts/test/Test4337ModuleAndHandler.sol#L53-L54):
```solidity
address payable safeAddress = payable(msg.sender);
ISafe safe = ISafe(safeAddress);
```

Does this need to be `payable`?

**Review by JohnGuilding — APPROVED** [review]
Looks good. Left a few non-blocking comments. Also got a small merge conflict

File: account-integrations/safe/src/utils/Safe4337Base.sol [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * In particular, this modifies eth-infinitism's BaseAccount to use
+ * `_msgSender()` when checking the call comes from the entry point, and pays
+ * the prefund using `execTransactionFromModule`.
+ */
+abstract contract Safe4337Base is BaseAccount, HandlerContext {
+    error NONCE_NOT_SEQUENTIAL();
+
+    function _requireFromEntryPoint() internal virtual view override {
+        require(
```
[the remark this answers is not in the document]
Comment (voltrevo) [reply account-integrations/safe/src/utils/Safe4337Base.sol:31]: Custom errors sound nice. Curious whether they have a good dev experience. If so I think that'd be a good improvement.

File: account-integrations/safe/src/utils/Safe4337Base.sol (hunk 1 above)
Comment (voltrevo) [reply account-integrations/safe/src/utils/Safe4337Base.sol:86]: Nope. `payable` is just a compile time type check thing. If the compiler is happy without `payable` (which it is), then it does the same thing.

File: account-integrations/safe/src/utils/Safe4337Base.sol [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            _msgSender() == address(entryPoint()),
+            "account: not from EntryPoint"
+        );
+    }
+
+    function _requireFromCurrentSafeOrEntryPoint() internal virtual view {
+        require(
+            _msgSender() == address(entryPoint()) ||
+            _msgSender() == address(_currentSafe()),
+            "account: not from EntryPoint"
```
[the remark this answers is not in the document]
Comment (voltrevo) [reply account-integrations/safe/src/utils/Safe4337Base.sol:41]: 👍 https://github.com/getwax/wax/pull/128

File: account-integrations/safe/src/utils/Safe4337Base.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (JohnGuilding) [reply account-integrations/safe/src/utils/Safe4337Base.sol:31]: Appears ethers has supported custom errors for a while https://blog.ricmoo.com/highlights-ethers-js-may-2021-2826e858277d. Does look like there can be some gotchas e.g. https://github.com/ethers-io/ethers.js/discussions/3291

Although optimisation probably isn't important at this stage of development, feels like an easy win for gas optimisation and is the direction a lot of projects seem to be going in e.g. the newly released Open Zeppelin V5 has replaced require statements with custom errors.

If we run into an issue decoding errors with the tools we're using, it looks like it isn't too difficult to decode them from the example given in this solidity blog post https://soliditylang.org/blog/2021/04/21/custom-errors/

**Pull request merged**